File: Leetcode/0879 Profitable Schemes.ts
const MOD = 1_000_000_007;

function profitableSchemes(n: number, minProfit: number, groups: number[], profits: number[]): number {
    const cache = new Array<number[][]>(profits.length);
    for (let i = 0; i < cache.length; i++) {
        cache[i] = new Array<number[]>(n + 1);
        for (let j = 0; j < cache[i].length; j++) {
            cache[i][j] = new Array<number>(minProfit + 1).fill(-1);
        }
    }
    function ways(i: number, groupSize: number, profit: number) {
        if (i === profits.length) {
            return profit >= minProfit ? 1 : 0;
        }
        if (cache[i][groupSize][profit] !== -1) {
            return cache[i][groupSize][profit];
        }
        let totalWays = ways(i + 1, groupSize, profit);
        if (groupSize + groups[i] <= n) {
            totalWays += ways(i + 1, groupSize + groups[i], Math.min(minProfit, profit + profits[i]));
        }
        return cache[i][groupSize][profit] = totalWays % MOD;
    }
    return ways(0, 0, 0);
}

File: Leetcode/0879 Profitable Schemes.py
MOD = int(1e9 + 7)

class Solution:
    def profitableSchemes(self, n, min_profit, groups, profits):
        cache = [[[None]*(min_profit + 1) for _ in range(n + 1) ] for _ in profits]
        
        def ways(i, group_size, profit):
            if i == len(profits):
                return 1 if profit >= min_profit else 0
            
            if cache[i][group_size][profit] is not None:
                return cache[i][group_size][profit]

            n_ways = ways(i + 1, group_size, profit)
            new_group_size = group_size + groups[i]
            if new_group_size <= n:
                n_ways += ways(i + 1, new_group_size, min(min_profit, profit + profits[i]))
            
            cache[i][group_size][profit] = n_ways % MOD
            return cache[i][group_size][profit]

        return ways(0, 0, 0)
